embulk-core/src/main/java/org/embulk/plugin/maven/MavenPluginSource.java

```diff
@@ -0,0 +1,168 @@
+    {
+        final String category;
+        if (InputPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "input";
+        } else if (OutputPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "output";
+        } else if (ParserPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "parser";
+        } else if (FormatterPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "formatter";
+        } else if (DecoderPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "decoder";
+        } else if (EncoderPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "encoder";
+        } else if (FilterPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "filter";
+        } else if (GuessPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "guess";
+        } else if (ExecutorPlugin.class.isAssignableFrom(pluginInterface)) {
+            category = "executor";
+        } else {
+            // unsupported plugin category
+            throw new PluginSourceNotMatchException("Plugin interface " + pluginInterface + " is not supported.");
+        }
+
+        if (pluginType.getSourceType() != PluginSource.Type.MAVEN) {
+            throw new PluginSourceNotMatchException();
+        }
+        final MavenPluginType mavenPluginType = (MavenPluginType) pluginType;
+
+        final PluginClassLoaderFactory pluginClassLoaderFactory =
+            this.injector.getInstance(PluginClassLoaderFactory.class);
+
+        final MavenArtifactFinder mavenArtifactFinder;
+        try {
+            mavenArtifactFinder = MavenArtifactFinder.create(getLocalMavenRepository());
+        }
+        catch (MavenRepositoryNotFoundException ex) {
+            throw new PluginSourceNotMatchException(ex);
+        }
+
+        final Path jarPath;
+        try {
+            jarPath = mavenArtifactFinder.findMavenArtifactJar(
+                mavenPluginType.getGroup(),
+                "embulk-" + category + "-" + mavenPluginType.getName(),
+                mavenPluginType.getVersion());
+        }
+        catch (MavenArtifactNotFoundException ex) {
+            throw new PluginSourceNotMatchException(ex);
+        }
+
+        final Class pluginMainClass;
+        try (JarPluginLoader loader = JarPluginLoader.load(jarPath, pluginClassLoaderFactory)) {
+            pluginMainClass = loader.getPluginMainClass();
+        }
+        catch (InvalidJarPluginException ex) {
+            throw new PluginSourceNotMatchException(ex);
+        }
+
+        if (!pluginInterface.isAssignableFrom(pluginMainClass)) {
+            throw new PluginSourceNotMatchException(
+                "Plugin class \"" + pluginMainClass.getName() + "\" is not a valid " + category + " plugin.");
+        }
+
+        final Object pluginMainObject;
+        try {
+            pluginMainObject = pluginMainClass.newInstance();
```

Comment:
Yeah, I now find that those `register` methods are:

* For `InjectedPluginSource` called from the initial Guice module:
    1. instantiating the plugin class,
    2. wrapping with `File*Runner` if needed, and
    3. registering for `InjectedPluginSource.PluginFactory` in  Guice so that `InjectedPluginSource#newPlugin` can find the plugin instance.
* For `JRubyPluginSource` called from `JRubyPluginSource#newPlugin`:
    1. instantiating the plugin class,
    2. wrapping with `File*Runner` if needed, and
    3. registering for `Embulk::PluginRegistry` in Ruby so that `JRubyPluginSource#newPlugin` can find the plugin instance.

In analogy with them, Maven-based plugin instances may need to be registered somewhere so that `MavenPluginSource#newPlugin` can find the instance.

But in case of `JRubyPluginSource`, that `register` is called from `newPlugin` to be used in `newPlugin`... "Registration" is done at the same time with `newPlugin`. Okay, therefore, it'd be okay to instantiate and wrap them with `File*Runner` in `MavenPluginSource#newPlugin`.

Sorry about that. I was also confused. Will be implementing that part...